src/instructlab/model/train.py

```diff
@@ -17,14 +24,97 @@
 logger = logging.getLogger(__name__)
 
 
+class TorchDeviceParam(click.ParamType):
+    """Parse and convert device string
+
+    Returns a torch.device object:
+    - type is one of 'cpu', 'cuda', 'hpu'
+    - index is None or device index (e.g. 0 for first GPU)
+    """
+
+    name = "deviceinfo"
+    supported_devices = {"cuda", "cpu", "hpu"}
+
+    def convert(self, value, param, ctx) -> "torch.device":
+        # pylint: disable=C0415
+        # Function local import, import torch can take more than a second
+        # Third Party
+        import torch
+
+        if not isinstance(value, torch.device):
+            try:
+                device = torch.device(value)
+            except RuntimeError as e:
+                self.fail(str(e), param, ctx)
+
+        if device.type not in self.supported_devices:
+            supported = ", ".join(repr(s) for s in sorted(self.supported_devices))
+            self.fail(
+                f"Unsupported device type '{device.type}'. Only devices "
+                f"types {supported}, and indexed device strings like 'cuda:0' "
+                "are supported for now.",
+                param,
+                ctx,
+            )
+
+        # Detect CUDA/ROCm device
+        if device.type == "cuda":
+            if not torch.cuda.is_available():
+                self.fail(
+                    f"{value}: Torch has no CUDA/ROCm support or could not detect "
```

Comment:
This is actually something that should be reported differently for merge.

We ought to say that CUDA isn't *available* rather than saying it's *unsupported*, the former is the canonical thing to report if the user's drivers or something are busted, the latter implies that we're using CUDA-less Torch, which we definitely aren't 😁